[cpp] Extern STL Containers and ArrayAccess

Aidan63 opened the pull request:
~~Adds generic externs for `std::vector` and `std::map` since other targets have externs for some of their native containers. I also added a new nativeArrayAccess meta which means extern classes implementing `ArrayAccess` will have square brackets generated for the array access in c++.~~ `ArrayAccess` was already supported but it assumed the extern class was of the `hx::Object` type and doesn't generate square bracket access in the c++. I didn't want to create a breaking change and there's no way of telling if an extern represents a non hxcpp object, so new meta it was.

~~I tried to enable `pf_overload` for c++ so I could add overloaded constructors on the externs but I ran into errors about core type constructors not matching on the ssl socket, not sure if anything can be done to fix that.~~

Aidan63 committed:
Add NativeArrayAccess meta for generating square bracket access.
Class must be extern and implement ArrayAccess

Simn commented:
Should these STL externs really be in std? I would think that a dedicated haxelib would be better for this.

Aidan63 commented:
That was my initial thought but since other targets have externs for their native containers I decided to put them up here. I can remove them though and just keep the ArrayAccess changes though.

Simn commented:
I thought other targets mostly only had that when the std implementation relied on it.

If there's no compelling reason to add these to std, I think a haxelib is the better choice for this. That also allows you to update things more easily.

Aidan63 commented:
Ah, you're probably right, I saw native containers in the c# and java package and assumed they were there for better interop. I've removed the stl externs and squashed the commits.

Simn commented:
Thanks!
